core/example/scafacos_example.cpp

```diff
@@ -0,0 +1,444 @@
+
+// User field enumeration. These will be used to index into the data set. Must
+// start at 0 and increment contiguously.
+//
+// NOTE: Users don't have to make this enum (or some other set of integral
+// constants) but it is a nice way to provide meaning to the different data
+// types and values assigned to the particles.
+//
+// NOTE: These enums are also ordered in the same way as the data in the
+// template parameters below.
+enum UserParticleFields
+{
+    PositionX = 0,
+    PositionY,
+    PositionZ,
+    Velocity,
+    Charge,
+    Potential,
+    Field,
+    Index
+};
+
+typedef struct 
+{
+    // MPI rank
+    int rank;
+    // number of MPI processes 
+    int n_proc;
+    // number of local particles
+    int n_local_particles;
+    // number of particles in each cartesian direction
+    // within the crystal
+    int loc_crystal[3]; 
+    // offset within the crystal
+    int off_crystal[3]; 
+    // cartesian dimensions of communicator
+    int dim[3];
+    // cartesian location of the local rank
+    int loc[3];
+    // communicator
+    MPI_Comm comm;
+} parallel_info;
+
+// Designate the types that the particles will hold.
+using ParticleDataTypes =
+    Cabana::MemberTypes<double,                       // (0) x-position type
+                        double,                       // (1) y-position type
+                        double,                       // (2) z-position type
+                        double[space_dim],            // (3) velocity type
+                        double,			              // (4) charge
+		                double,			              // (5) potential
+		                double[space_dim],		      // (6) electric field values
+                        int                           // (7) global index 
+                        >;
+
+// Declare the memory space.
+using MemorySpace = Cabana::HostSpace;
+
+// Declare the length of the internal vectors
+const int VectorLength = 8;
+
+// Set the type for the particle AoSoA.
+using ParticleList = Cabana::AoSoA<ParticleDataTypes,MemorySpace,VectorLength>;
+
+//---------------------------------------------------------------------------//
+// Helper functions.
+//---------------------------------------------------------------------------//
+// Function to intitialize the particles.
+void initializeParticles( ParticleList& particles, 
+                          double gap_space, 
+                          parallel_info& info )
+{
+    auto p_x = particles.slice<PositionX>();
+    auto p_y = particles.slice<PositionY>();
+    auto p_z = particles.slice<PositionZ>();
+    auto v = particles.slice<Velocity>();
+    auto q = particles.slice<Charge>();
+    auto pot = particles.slice<Potential>();
+    auto field = particles.slice<Field>();
+    auto indx = particles.slice<Index>();
+
+    int offset;
+
+    //compute offset with MPI_Exscan
+    MPI_Exscan(&(info.n_local_particles), &offset, 1, MPI_INT, MPI_SUM, info.comm);
+    if (info.rank == 0) offset = 0;
```

Comment:
Maybe add a comment here that the MPI standard says that the receive buffer on rank 0 is undefined.